xcm/xcm-simulator/src/lib.rs

```diff
@@ -32,13 +29,30 @@ pub use polkadot_runtime_parachains::{
 	dmp,
 	ump::{self, MessageId, UmpSink, XcmSink},
 };
+pub use sp_io::TestExternalities;
+pub use sp_std::{cell::RefCell, collections::vec_deque::VecDeque, marker::PhantomData};
 pub use xcm::{v0::prelude::*, VersionedXcm};
 pub use xcm_executor::XcmExecutor;
 
 pub trait TestExt {
+	/// Initialize the test environment.
 	fn new_ext() -> sp_io::TestExternalities;
+	/// Resets the state of the test environment.
 	fn reset_ext();
-	fn execute_with<R>(execute: impl FnOnce() -> R) -> R;
+	/// Execute code in the context of the test externalities, without automatic
+	/// message processing. All messages in the message buses can be processed
+	/// by calling `Self::dispatch_xcm_queue()`.
+	fn execute_and_queue_xcm<R>(execute: impl FnOnce() -> R) -> R;
+	/// Process all messages in the message buses
+	fn dispatch_xcm_queue();
+	/// Execute some code in the context of the test externalities, with
+	/// automatic message processing.
+	/// Messages are dispatched once the passed closure completes.
+	fn execute_and_dispatch_xcm<R>(execute: impl FnOnce() -> R) -> R {
```

Comment:
I would prefer the previous fn name `execute_with`(and `execute_with_no_dispatch`). In real life networks, messages are dispatched automatically after sending extrinsics, which is the same behavior with here, so there is no need to explicitly say it would dispatch xcm. Also, `execute_with` is same to `TestExternalities`'s API and it's intuitive to use.

Reply:
I agree. In general, the common path should have short names and the uncommon path can have longer names.

There was some discussion over the names so I changed them, but I prefer your suggestion (changing them back now).